README.md
# dure README

This is the README for your extension "dure". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------
## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "dure" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import { commands, window, ExtensionContext, Uri, workspace, TreeItem, TreeItemCollapsibleState, Disposable, TreeDataProvider, ThemeIcon, ProgressLocation, FileDecoration, EventEmitter, Event, Command } from 'vscode';

import { pathConfig } from './configuration';

// import path from "path"

// ,
// 		"snippets": [
// 			{
// 				"language": "cpp",
// 				"path": "./snippets/snippets.json"
// 			}
// 		]

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate({ subscriptions }: ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	// console.log('Congratulations, your extension "dure" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	// let disposable = vscode.commands.registerCommand('dure.helloWorld', () => {
	// 	// The code you place here will be executed every time your command is executed
	// 	// Display a message box to the user
	// 	vscode.window.showInformationMessage('Hello World from CodeReplace!');
	// });

	const previewWelcome = commands.registerCommand('dure.welcome', () => {
		// console.log(commands.executeCommand('git.showOutput'));
		// console.log(commands.executeCommand('git.fetch'));
		// commands.executeCommand("git.addRemote")
		// 	.then((result) => {
		// 		console.log(result);
		// 	});
		// window.showInformationMessage('This is the message', {
		// 	modal: true,
		// 	detail: 'This is the detail. Rendered less prominent, but with more space for, well, details.'
		// });
		let pathKeys = Object.keys(pathConfig);
		window.showQuickPick(pathKeys.map(key => ({
			label: key,
			description: "current source",
			detail: "current source",
		})), {
			title: "选择源仓库",
			placeHolder: "用于同步代码的仓库",
			ignoreFocusOut: true,
			canPickMany: false,
			onDidSelectItem(key) {
				console.dir(key);
			}
		});

	});

	class SyncFile extends TreeItem {
		// iconPath = {
		// 	light: path.join(__filename, '..', '..', 'resources', 'light', 'dependency.svg'),
		// 	dark: path.join(__filename, '..', '..', 'resources', 'dark', 'dependency.svg')
		// };

		constructor(
			public label: string,
			public readonly collapsibleState: TreeItemCollapsibleState,
			public iconPath: ThemeIcon,
			public readonly command: Command,
			public contextValue: string | undefined,

		) {
			super({ label, highlights: [[0, label.length]] }, collapsibleState);

			this.tooltip = label;
			this.description = label;
			this.contextValue = contextValue;
			this.iconPath = iconPath;
			this.command = command;
		}

	}

	let bool = (): boolean => Math.random() > 0.5;

	let generatorRandomTestData = (index: number, icon: string = "circle-large-outline") => new SyncFile(
		`测试${Math.ceil(Math.random() * 1000)}`,
		bool() ? 1 : 0,
		new ThemeIcon(icon),
		{ title: "asd", command: "dure.addSyncFile", arguments: [index] },
		`${bool() ? 1 : 0}`
	);

	let _fileData = (): SyncFile[] => Array.from({ length: 0 }, (el, i): SyncFile => generatorRandomTestData(i));

	let generateData: SyncFile[] = _fileData();

	// 	{ name: `测试${Math.ceil(Math.random() * 1000)}`, index: 1, icon: "circle-large-outline" },

	class FileTreeData {
		private _onDidChangeTreeData: EventEmitter<TreeItem | undefined | void> = new EventEmitter<TreeItem | undefined | void>();
		readonly onDidChangeTreeData: Event<TreeItem | undefined | void> = this._onDidChangeTreeData.event;

		constructor() { }

		refresh(): void {
			// let treeItem = new TreeItem({ label: "撒大声地", highlights: [[0, 6]] }, 1);
			// treeItem.iconPath = new ThemeIcon('pass-filled');
			// // treeItem.iconPath = Uri.joinPath(Uri.parse(".."), 'resources/sync.svg');
			// // treeItem.contextValue = "asdasdas";
			// treeItem.command = {
			// 	title: "+++++++++",
			// 	command: "dure.addSyncFile",
			// 	arguments: [this.refresh]
			// };
			// this.onDidChangeTreeData((value) => {
			// 	console.dir(value);
			// 	console.dir('^^^^^^^^^^^^^^^^^^^');
			// 	return treeItem;
			// });

			this._onDidChangeTreeData.fire();
		}

		getChildren(el: SyncFile): SyncFile[] {
			console.log(el);
			console.log("0000000000000000000000");
			// return Array.from({ length: 6 }, (el, i) => ({ name: `测试${Math.ceil(Math.random() * 1000)}`, index: 1, icon: "circle-large-outline" }));
			return !el ? generateData : [];
		}

		getTreeItem(item: SyncFile) {
			// console.dir(item);
			// console.dir('99999999999999');
			// let treeItem = new TreeItem({ label: item.name, highlights: [[0, item.name.length]] }, item.index);
			// treeItem.iconPath = new ThemeIcon(item.icon);
			// // treeItem.iconPath = Uri.joinPath(Uri.parse(".."), 'resources/sync.svg');
			// // treeItem.contextValue = "asdasdas";
			// treeItem.contextValue = item.index;
			// treeItem.tooltip = "asfasfasfa";
			// treeItem.command = {
			// 	title: "asd",
			// 	command: "dure.addSyncFile",
			// 	arguments: [this.refresh]
			// };
			return item;
		}

		// getParent() {
		// 	console.dir("**************");
		// }
	}

	let treeData = new FileTreeData();

	treeData.onDidChangeTreeData((data: any) => {
		console.dir('09090909090');
		console.dir(data);
	});

	window.registerTreeDataProvider("sync-files", treeData);

	commands.registerCommand('dure.addSyncFile', (index) => {
		console.dir('--------------');
		// console.dir(el);
		// console.dir(el2);
		// el.label = { label: "asdasd" };
		// el.iconPath = new ThemeIcon('pass-filled');
		// pass
		// _fileData[3].name = "啊实打实地方";
		// _fileData[3].icon = "pass-filled";
		// refresh();
		generateData.splice(index, 1, generatorRandomTestData(index, "pass-filled"));
		treeData.refresh();
		console.dir("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
	});
	// commands.registerCommand('dure.addSyncFile', (treeData, el: TreeItem, el2) => {
	// 	treeData.fi
	// });

	const test = commands.registerCommand('dure.test', () => {
		// console.dir(Uri.joinPath(Uri.parse(".."), 'resources/sync.svg'));
		// console.dir(Uri.parse('..'));
		// console.dir(__dirname);
		// console.dir(__filename);
		let treeData: TreeDataProvider<any> = {
			onDidChangeTreeData: new EventEmitter().event,
			getChildren(el) {
				// console.log(el);
				console.log("0000000000000000000000");
				return Array.from({ length: 6 }, (el, i) => ({ name: `测试${Math.ceil(Math.random() * 1000)}`, index: 1, icon: "circle-large-outline" }));
			},
			getTreeItem(el: Record<string, any>) {
				// console.dir(el);
				console.dir('99999999999999');
				let treeItem = new TreeItem({ label: `测试${Math.ceil(Math.random() * 1000)}`, highlights: [[0, el.name.length]] }, el.index)
				treeItem.iconPath = new ThemeIcon("circle-large-outline");
				// treeItem.iconPath = Uri.joinPath(Uri.parse(".."), 'resources/sync.svg');
				// treeItem.contextValue = "asdasdas";
				treeItem.tooltip = "asfasfasfa";
				treeItem.command = {
					title: "asd",
					command: "dure.addSyncFile",
					arguments: [this.onDidChangeTreeData, treeItem, "asdasd"]
				};
				treeItem.accessibilityInformation = {
					label: "asdasdas"
				};
				return treeItem;
			},
			getParent(el) {
				console.dir("**************");
			},
			// resolveTreeItem(item) {
			// 	item.iconPath = new ThemeIcon('pass-filled');
			// 	console.log('+++++++++++++++');
			// 	console.log(item);
			// 	return item;
			// }
		};
		// treeData.onDidChangeTreeData(() => { });
		let tree = window.createTreeView("sync-files", {
			showCollapseAll: true,
			canSelectMany: true,
			treeDataProvider: treeData
		});

		tree.onDidChangeSelection((el) => {
			console.dir('onDidChangeSelection');
			console.dir(el);
		}, null);
		tree.onDidExpandElement((el) => {
			console.dir('onDidExpandElement');
			console.dir(el);
		});
		tree.onDidCollapseElement((el) => {
			console.dir('onDidCollapseElement');
			console.dir(el);
		});
	});

	subscriptions.push(test);
	subscriptions.push(previewWelcome);

	// window.withProgress({
	// 	location: ProgressLocation.Notification,
	// 	title: "I am long running!",
	// 	cancellable: true
	// }, (progress, token) => {
	// 	token.onCancellationRequested(() => {
	// 		console.log("User canceled the long running operation");
	// 	});

	// 	progress.report({ increment: 0 });

	// 	setTimeout(() => {
	// 		progress.report({ increment: 10, message: "I am long running! - still going..." });
	// 	}, 1000);

	// 	setTimeout(() => {
	// 		progress.report({ increment: 40, message: "I am long running! - still going even more..." });
	// 	}, 2000);

	// 	setTimeout(() => {
	// 		progress.report({ increment: 50, message: "I am long running! - almost there..." });
	// 	}, 3000);

	// 	const p = new Promise<void>(resolve => {
	// 		setTimeout(() => {
	// 			resolve();
	// 		}, 5000);
	// 	});

	// 	return p;
	// });

	// console.log('*****************');
	// console.log(workspace.workspaceFolders);
}

// this method is called when your extension is deactivated
export function deactivate() { }

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode_1 = require("vscode");
const configuration_1 = require("./configuration");
// import path from "path"
// ,
// 		"snippets": [
// 			{
// 				"language": "cpp",
// 				"path": "./snippets/snippets.json"
// 			}
// 		]
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate({ subscriptions }) {
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    // console.log('Congratulations, your extension "dure" is now active!');
    // The command has been defined in the package.json file
    // Now provide the implementation of the command with registerCommand
    // The commandId parameter must match the command field in package.json
    // let disposable = vscode.commands.registerCommand('dure.helloWorld', () => {
    // 	// The code you place here will be executed every time your command is executed
    // 	// Display a message box to the user
    // 	vscode.window.showInformationMessage('Hello World from CodeReplace!');
    // });
    const previewWelcome = vscode_1.commands.registerCommand('dure.welcome', () => {
        // console.log(commands.executeCommand('git.showOutput'));
        // console.log(commands.executeCommand('git.fetch'));
        // commands.executeCommand("git.addRemote")
        // 	.then((result) => {
        // 		console.log(result);
        // 	});
        // window.showInformationMessage('This is the message', {
        // 	modal: true,
        // 	detail: 'This is the detail. Rendered less prominent, but with more space for, well, details.'
        // });
        let pathKeys = Object.keys(configuration_1.pathConfig);
        vscode_1.window.showQuickPick(pathKeys.map(key => ({
            label: key,
            description: "current source",
            detail: "current source",
        })), {
            title: "选择源仓库",
            placeHolder: "用于同步代码的仓库",
            ignoreFocusOut: true,
            canPickMany: false,
            onDidSelectItem(key) {
                console.dir(key);
            }
        });
    });
    class SyncFile extends vscode_1.TreeItem {
        // iconPath = {
        // 	light: path.join(__filename, '..', '..', 'resources', 'light', 'dependency.svg'),
        // 	dark: path.join(__filename, '..', '..', 'resources', 'dark', 'dependency.svg')
        // };
        constructor(label, collapsibleState, iconPath, command, contextValue) {
            super({ label, highlights: [[0, label.length]] }, collapsibleState);
            this.label = label;
            this.collapsibleState = collapsibleState;
            this.iconPath = iconPath;
            this.command = command;
            this.contextValue = contextValue;
            this.tooltip = label;
            this.description = label;
            this.contextValue = contextValue;
            this.iconPath = iconPath;
            this.command = command;
        }
    }
    let bool = () => Math.random() > 0.5;
    let generatorRandomTestData = (index, icon = "circle-large-outline") => new SyncFile(`测试${Math.ceil(Math.random() * 1000)}`, bool() ? 1 : 0, new vscode_1.ThemeIcon(icon), { title: "asd", command: "dure.addSyncFile", arguments: [index] }, `${bool() ? 1 : 0}`);
    let _fileData = () => Array.from({ length: 0 }, (el, i) => generatorRandomTestData(i));
    let generateData = _fileData();
    // 	{ name: `测试${Math.ceil(Math.random() * 1000)}`, index: 1, icon: "circle-large-outline" },
    class FileTreeData {
        constructor() {
            this._onDidChangeTreeData = new vscode_1.EventEmitter();
            this.onDidChangeTreeData = this._onDidChangeTreeData.event;
        }
        refresh() {
            // let treeItem = new TreeItem({ label: "撒大声地", highlights: [[0, 6]] }, 1);
            // treeItem.iconPath = new ThemeIcon('pass-filled');
            // // treeItem.iconPath = Uri.joinPath(Uri.parse(".."), 'resources/sync.svg');
            // // treeItem.contextValue = "asdasdas";
            // treeItem.command = {
            // 	title: "+++++++++",
            // 	command: "dure.addSyncFile",
            // 	arguments: [this.refresh]
            // };
            // this.onDidChangeTreeData((value) => {
            // 	console.dir(value);
            // 	console.dir('^^^^^^^^^^^^^^^^^^^');
            // 	return treeItem;
            // });
            this._onDidChangeTreeData.fire();
        }
        getChildren(el) {
            console.log(el);
            console.log("0000000000000000000000");
            // return Array.from({ length: 6 }, (el, i) => ({ name: `测试${Math.ceil(Math.random() * 1000)}`, index: 1, icon: "circle-large-outline" }));
            return !el ? generateData : [];
        }
        getTreeItem(item) {
            // console.dir(item);
            // console.dir('99999999999999');
            // let treeItem = new TreeItem({ label: item.name, highlights: [[0, item.name.length]] }, item.index);
            // treeItem.iconPath = new ThemeIcon(item.icon);
            // // treeItem.iconPath = Uri.joinPath(Uri.parse(".."), 'resources/sync.svg');
            // // treeItem.contextValue = "asdasdas";
            // treeItem.contextValue = item.index;
            // treeItem.tooltip = "asfasfasfa";
            // treeItem.command = {
            // 	title: "asd",
            // 	command: "dure.addSyncFile",
            // 	arguments: [this.refresh]
            // };
            return item;
        }
    }
    let treeData = new FileTreeData();
    treeData.onDidChangeTreeData((data) => {
        console.dir('09090909090');
        console.dir(data);
    });
    vscode_1.window.registerTreeDataProvider("sync-files", treeData);
    vscode_1.commands.registerCommand('dure.addSyncFile', (index) => {
        console.dir('--------------');
        // console.dir(el);
        // console.dir(el2);
        // el.label = { label: "asdasd" };
        // el.iconPath = new ThemeIcon('pass-filled');
        // pass
        // _fileData[3].name = "啊实打实地方";
        // _fileData[3].icon = "pass-filled";
        // refresh();
        generateData.splice(index, 1, generatorRandomTestData(index, "pass-filled"));
        treeData.refresh();
        console.dir("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
    });
    // commands.registerCommand('dure.addSyncFile', (treeData, el: TreeItem, el2) => {
    // 	treeData.fi
    // });
    const test = vscode_1.commands.registerCommand('dure.test', () => {
        // console.dir(Uri.joinPath(Uri.parse(".."), 'resources/sync.svg'));
        // console.dir(Uri.parse('..'));
        // console.dir(__dirname);
        // console.dir(__filename);
        let treeData = {
            onDidChangeTreeData: new vscode_1.EventEmitter().event,
            getChildren(el) {
                // console.log(el);
                console.log("0000000000000000000000");
                return Array.from({ length: 6 }, (el, i) => ({ name: `测试${Math.ceil(Math.random() * 1000)}`, index: 1, icon: "circle-large-outline" }));
            },
            getTreeItem(el) {
                // console.dir(el);
                console.dir('99999999999999');
                let treeItem = new vscode_1.TreeItem({ label: `测试${Math.ceil(Math.random() * 1000)}`, highlights: [[0, el.name.length]] }, el.index);
                treeItem.iconPath = new vscode_1.ThemeIcon("circle-large-outline");
                // treeItem.iconPath = Uri.joinPath(Uri.parse(".."), 'resources/sync.svg');
                // treeItem.contextValue = "asdasdas";
                treeItem.tooltip = "asfasfasfa";
                treeItem.command = {
                    title: "asd",
                    command: "dure.addSyncFile",
                    arguments: [this.onDidChangeTreeData, treeItem, "asdasd"]
                };
                treeItem.accessibilityInformation = {
                    label: "asdasdas"
                };
                return treeItem;
            },
            getParent(el) {
                console.dir("**************");
            },
            // resolveTreeItem(item) {
            // 	item.iconPath = new ThemeIcon('pass-filled');
            // 	console.log('+++++++++++++++');
            // 	console.log(item);
            // 	return item;
            // }
        };
        // treeData.onDidChangeTreeData(() => { });
        let tree = vscode_1.window.createTreeView("sync-files", {
            showCollapseAll: true,
            canSelectMany: true,
            treeDataProvider: treeData
        });
        tree.onDidChangeSelection((el) => {
            console.dir('onDidChangeSelection');
            console.dir(el);
        }, null);
        tree.onDidExpandElement((el) => {
            console.dir('onDidExpandElement');
            console.dir(el);
        });
        tree.onDidCollapseElement((el) => {
            console.dir('onDidCollapseElement');
            console.dir(el);
        });
    });
    subscriptions.push(test);
    subscriptions.push(previewWelcome);
    // window.withProgress({
    // 	location: ProgressLocation.Notification,
    // 	title: "I am long running!",
    // 	cancellable: true
    // }, (progress, token) => {
    // 	token.onCancellationRequested(() => {
    // 		console.log("User canceled the long running operation");
    // 	});
    // 	progress.report({ increment: 0 });
    // 	setTimeout(() => {
    // 		progress.report({ increment: 10, message: "I am long running! - still going..." });
    // 	}, 1000);
    // 	setTimeout(() => {
    // 		progress.report({ increment: 40, message: "I am long running! - still going even more..." });
    // 	}, 2000);
    // 	setTimeout(() => {
    // 		progress.report({ increment: 50, message: "I am long running! - almost there..." });
    // 	}, 3000);
    // 	const p = new Promise<void>(resolve => {
    // 		setTimeout(() => {
    // 			resolve();
    // 		}, 5000);
    // 	});
    // 	return p;
    // });
    // console.log('*****************');
    // console.log(workspace.workspaceFolders);
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map